<a href="https://colab.research.google.com/github/PRITESHDAYAT/Priteshfiledemo/blob/main/EDA_NYC_Taxi_Analysis_Pritesh.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **New York City Yellow Taxi Data**

## Objective
In this case study you will be learning exploratory data analysis (EDA) with the help of a dataset on yellow taxi rides in New York City. This will enable you to understand why EDA is an important step in the process of data science and machine learning.

## **Problem Statement**
As an analyst at an upcoming taxi operation in NYC, you are tasked to use the 2023 taxi trip data to uncover insights that could help optimise taxi operations. The goal is to analyse patterns in the data that can inform strategic decisions to improve service efficiency, maximise revenue, and enhance passenger experience.

## Tasks
You need to perform the following steps for successfully completing this assignment:
1. Data Loading
2. Data Cleaning
3. Exploratory Analysis: Bivariate and Multivariate
4. Creating Visualisations to Support the Analysis
5. Deriving Insights and Stating Conclusions

---

**NOTE:** The marks given along with headings and sub-headings are cumulative marks for those particular headings/sub-headings.<br>

The actual marks for each task are specified within the tasks themselves.

For example, marks given with heading *2* or sub-heading *2.1* are the cumulative marks, for your reference only. <br>

The marks you will receive for completing tasks are given with the tasks.

Suppose the marks for two tasks are: 3 marks for 2.1.1 and 2 marks for 3.2.2, or
* 2.1.1 [3 marks]
* 3.2.2 [2 marks]

then, you will earn 3 marks for completing task 2.1.1 and 2 marks for completing task 3.2.2.


---

## Data Understanding
The yellow taxi trip records include fields capturing pick-up and drop-off dates/times, pick-up and drop-off locations, trip distances, itemized fares, rate types, payment types, and driver-reported passenger counts.

The data is stored in Parquet format (*.parquet*). The dataset is from 2009 to 2024. However, for this assignment, we will only be using the data from 2023.

The data for each month is present in a different parquet file. You will get twelve files for each of the months in 2023.

The data was collected and provided to the NYC Taxi and Limousine Commission (TLC) by technology providers like vendors and taxi hailing apps. <br>

You can find the link to the TLC trip records page here: https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page

w3###  Data Description
You can find the data description here: [Data Dictionary](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf)

**Trip Records**



|Field Name       |description |
|:----------------|:-----------|
| VendorID | A code indicating the TPEP provider that provided the record. <br> 1= Creative Mobile Technologies, LLC; <br> 2= VeriFone Inc. |
| tpep_pickup_datetime | The date and time when the meter was engaged.  |
| tpep_dropoff_datetime | The date and time when the meter was disengaged.   |
| Passenger_count | The number of passengers in the vehicle. <br> This is a driver-entered value. |
| Trip_distance | The elapsed trip distance in miles reported by the taximeter. |
| PULocationID | TLC Taxi Zone in which the taximeter was engaged |
| DOLocationID | TLC Taxi Zone in which the taximeter was disengaged |
|RateCodeID |The final rate code in effect at the end of the trip.<br> 1 = Standard rate <br> 2 = JFK <br> 3 = Newark <br>4 = Nassau or Westchester <br>5 = Negotiated fare <br>6 = Group ride |
|Store_and_fwd_flag |This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.  <br>Y= store and forward trip <br>N= not a store and forward trip |
|Payment_type| A numeric code signifying how the passenger paid for the trip. <br> 1 = Credit card <br>2 = Cash <br>3 = No charge <br>4 = Dispute <br>5 = Unknown <br>6 = Voided trip |
|Fare_amount| The time-and-distance fare calculated by the meter. <br>Extra Miscellaneous extras and surcharges.  Currently, this only includes the 0.50 and 1 USD rush hour and overnight charges. |
|MTA_tax |0.50 USD MTA tax that is automatically triggered based on the metered rate in use. |
|Improvement_surcharge | 0.30 USD improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. |
|Tip_amount |Tip amount – This field is automatically populated for credit card tips. Cash tips are not included. |
| Tolls_amount | Total amount of all tolls paid in trip.  |
| total_amount | The total amount charged to passengers. Does not include cash tips. |
|Congestion_Surcharge |Total amount collected in trip for NYS congestion surcharge. |
| Airport_fee | 1.25 USD for pick up only at LaGuardia and John F. Kennedy Airports|

Although the amounts of extra charges and taxes applied are specified in the data dictionary, you will see that some cases have different values of these charges in the actual data.

**Taxi Zones**

Each of the trip records contains a field corresponding to the location of the pickup or drop-off of the trip, populated by numbers ranging from 1-263.

These numbers correspond to taxi zones, which may be downloaded as a table or map/shapefile and matched to the trip records using a join.

This is covered in more detail in later sections.

---

## **1** Data Preparation

<font color = red>[5 marks]</font> <br>

### Import Libraries

In [9]:
# Import warnings
import warnings
warnings.filterwarnings('ignore') #“Hide all warning messages”


In [10]:
# Import the libraries you will be using for analysis
import numpy as np              # Numerical operations
import pandas as pd             # Data handling and analysis
import matplotlib.pyplot as plt # Data visualization
import seaborn as sns            # Advanced visualizations
import statistics as stats       # Statistical calculations


In [11]:
#Recommended versions
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2

# Check versions
print("numpy version:", np.__version__)
print("pandas version:", pd.__version__)
print("matplotlib version:", plt.matplotlib.__version__)
print("seaborn version:", sns.__version__)

numpy version: 2.0.2
pandas version: 2.2.2
matplotlib version: 3.10.0
seaborn version: 0.13.2


### **1.1** Load the dataset
<font color = red>[5 marks]</font> <br>

You will see twelve files, one for each month.

To read parquet files with Pandas, you have to follow a similar syntax as that for CSV files.

`df = pd.read_parquet('file.parquet')`

In [18]:
# Try loading one file
df_2023_1 = pd.read_parquet('/content/2023-1.parquet')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

How many rows are there? Do you think handling such a large number of rows is computationally feasible when we have to combine the data for all twelve months into one?

To handle this, we need to sample a fraction of data from each of the files. How to go about that? Think of a way to select only some portion of the data from each month's file that accurately represents the trends.

#### Sampling the Data
> One way is to take a small percentage of entries for pickup in every hour of a date. So, for all the days in a month, we can iterate through the hours and select 5% values randomly from those. Use `tpep_pickup_datetime` for this. Separate date and hour from the datetime values and then for each date, select some fraction of trips for each of the 24 hours.

To sample data, you can use the `sample()` method. Follow this syntax:

```Python
# sampled_data is an empty DF to keep appending sampled data of each hour
# hour_data is the DF of entries for an hour 'X' on a date 'Y'

sample = hour_data.sample(frac = 0.05, random_state = 42)
# sample 0.05 of the hour_data
# random_state is just a seed for sampling, you can define it yourself

sampled_data = pd.concat([sampled_data, sample]) # adding data for this hour to the DF
```

This *sampled_data* will contain 5% values selected at random from each hour.

Note that the code given above is only the part that will be used for sampling and not the complete code required for sampling and combining the data files.

Keep in mind that you sample by date AND hour, not just hour. (Why?)

In [68]:
df_2023_1 = pd.read_parquet('/content/2023-1.parquet')  # 1
df.info()
df_2023_1['hour'] = df_2023_1["tpep_pickup_datetime"].dt.hour # remove the one hour from df_2023_1_Parquet file
sample_df_2023_1 = df_2023_1.groupby("hour").sample(frac=0.05, random_state=42) # give 5 hour of each start day
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023_1.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023_1.info()
sample_df_2023_1.head()


<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
784218          2  2023-01-10 00:44:09   2023-01-10 00:51:23              1.0   
1316921         1  2023-01-15 00:32:10   2023-01-15 00:42:27              1.0   
2732390         1  2023-01-29 00:29:35   2023-01-29 00:42:50              2.0   
1203249         2  2023-01-14 00:14:30   2023-01-14 00:27:04              1.0   
137926          2  2023-01-03 00:55:05   2023-01-03 01:03:28              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
784218            1.97         1.0                  N           142   
1316921           1.50         1.0                  N           113   
2732390           2.60         1.0                  N           263   
1203249           1.79         1.0                  N            79   
137926            3.34         1.0                  N           132   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
784218            164             1         10.7    1.0      0.5        3.14   
1316921            79             1         10.0    3.5      0.5        2.00   
2732390           143             1         15.6    3.5      0.5        4.10   
1203249           158             1         13.5    1.0      0.5        1.48   
137926            132             2         15.6    1.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
784218            0.0                    1.0         18.84   
1316921           0.0                    1.0         17.00   
2732390           0.0                    1.0         24.70   
1203249           0.0                    1.0         19.98   
137926            0.0                    1.0         19.35   

         congestion_surcharge  airport_fee  hour  
784218                    2.5         0.00     0  
1316921                   2.5         0.00     0  
2732390                   2.5         0.00     0  
1203249                   2.5         0.00     0  
137926                    0.0         1.25     0

In [56]:
df_2023_2 = pd.read_parquet('/content/2023-2.parquet')  # 2
df.info()
df_2023_2['hour'] = df_2023_2["tpep_pickup_datetime"].dt.hour # remove the one hour from df_2023_1_Parquet file
sample_df_2023 = df_2023_2.groupby("hour").sample(frac=0.05, random_state=42) # give 5 hour of each start  ( making a new file)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()


<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1621934         2  2023-03-16 00:30:35   2023-03-16 00:35:14              1.0   
984044          2  2023-03-10 00:31:58   2023-03-10 00:38:57              1.0   
2257934         1  2023-03-22 00:16:12   2023-03-22 00:31:27              2.0   
1972252         2  2023-03-19 00:21:13   2023-03-19 00:37:16              2.0   
2693913         2  2023-03-26 00:29:25   2023-03-26 00:56:03              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1621934           0.85         1.0                  N            41   
984044            1.50         1.0                  N            79   
2257934           2.30         1.0                  N           114   
1972252           2.17         1.0                  N            90   
2693913           5.90         1.0                  N           114   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1621934            75             1          7.2    1.0      0.5        1.94   
984044            234             1          9.3    1.0      0.5        2.86   
2257934           231             1         14.2    3.5      0.5        3.80   
1972252           231             1         16.3    1.0      0.5        3.20   
2693913            37             1         31.0    1.0      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
1621934           0.0                    1.0         11.64   
984044            0.0                    1.0         17.16   
2257934           0.0                    1.0         23.00   
1972252           0.0                    1.0         24.50   
2693913           0.0                    1.0         38.00   

         congestion_surcharge  Airport_fee  hour  
1621934                   0.0          0.0     0  
984044                    2.5          0.0     0  
2257934                   2.5          0.0     0  
1972252                   2.5          0.0     0  
2693913                   2.5          0.0     0

In [57]:
df_2023_3 = pd.read_parquet('/content/2023-3.parquet')  # 3
df.info()
df_2023_3['hour'] = df_2023_3["tpep_pickup_datetime"].dt.hour # remove the one hour from df_2023_1_Parquet file
sample_df_2023 = df_2023_3.groupby("hour").sample(frac=0.05, random_state=42) # give 5 hour of each start day
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
547140          2  2023-06-06 00:21:40   2023-06-06 00:29:16              2.0   
2912863         2  2023-06-28 00:58:26   2023-06-28 01:03:56              3.0   
446923          2  2023-06-05 00:27:28   2023-06-05 00:31:55              1.0   
1682438         2  2023-06-16 00:26:38   2023-06-16 00:32:46              1.0   
665289          1  2023-06-07 00:13:44   2023-06-07 00:26:51              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
547140            1.98         1.0                  N           230   
2912863           1.32         1.0                  N            24   
446923            0.89         1.0                  N           246   
1682438           1.23         1.0                  N           161   
665289            4.30         1.0                  N            79   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
547140            239             1         10.7    1.0      0.5        3.14   
2912863           238             1          8.6    1.0      0.5        2.72   
446923             90             2          6.5    1.0      0.5        0.00   
1682438           237             1          9.3    1.0      0.5        2.86   
665289            263             1         18.4    3.5      0.5        4.65   

         tolls_amount  improvement_surcharge  total_amount  \
547140            0.0                    1.0         18.84   
2912863           0.0                    1.0         16.32   
446923            0.0                    1.0         11.50   
1682438           0.0                    1.0         17.16   
665289            0.0                    1.0         28.05   

         congestion_surcharge  Airport_fee  hour  
547140                    2.5          0.0     0  
2912863                   2.5          0.0     0  
446923                    2.5          0.0     0  
1682438                   2.5          0.0     0  
665289                    2.5          0.0     0

In [59]:
df_2023_4 = pd.read_parquet('/content/2023-4.parquet')  # 4
df.info()
df_2023_4['hour'] = df_2023_4["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_4.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2753974         2  2023-08-07 00:42:00   2023-08-07 00:54:00              NaN   
2228375         2  2023-08-26 00:53:41   2023-08-26 01:27:29              3.0   
1260372         2  2023-08-15 00:30:17   2023-08-15 00:37:16              2.0   
475920          2  2023-08-06 00:10:10   2023-08-06 00:19:09              1.0   
2648741         2  2023-08-31 00:40:26   2023-08-31 00:50:07              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
2753974           2.58         NaN               None           186   
2228375          19.31         2.0                  N           132   
1260372           1.35         1.0                  N           209   
475920            1.48         1.0                  N           234   
2648741           3.01         1.0                  N            48   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
2753974           125             0        28.73    0.0      0.5        6.55   
2228375           236             1        70.00    0.0      0.5       16.19   
1260372           148             1        10.00    1.0      0.5        3.00   
475920            148             1        10.70    1.0      0.5        3.14   
2648741           236             1        14.90    1.0      0.5        3.98   

         tolls_amount  improvement_surcharge  total_amount  \
2753974          0.00                    1.0         39.28   
2228375          6.94                    1.0         98.88   
1260372          0.00                    1.0         18.00   
475920           0.00                    1.0         18.84   
2648741          0.00                    1.0         23.88   

         congestion_surcharge  Airport_fee  hour  
2753974                   NaN          NaN     0  
2228375                   2.5         1.75     0  
1260372                   2.5         0.00     0  
475920                    2.5         0.00     0  
2648741                   2.5         0.00     0

In [60]:
df_2023_5 = pd.read_parquet('/content/2023-5.parquet')  # 5
df.info()
df_2023_5['hour'] = df_2023_5["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_5.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2450651         2  2023-02-25 00:15:02   2023-02-25 00:24:27              6.0   
2019149         2  2023-02-21 00:13:13   2023-02-21 00:38:54              1.0   
326721          2  2023-02-04 00:16:31   2023-02-04 00:27:47              1.0   
1753508         2  2023-02-18 00:41:05   2023-02-18 01:02:56              1.0   
2562731         2  2023-02-26 00:02:45   2023-02-26 00:22:20              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
2450651           1.03         1.0                  N           249   
2019149          15.16         1.0                  N           132   
326721            5.15         1.0                  N           164   
1753508           4.99         1.0                  N            79   
2562731           4.13         1.0                  N           234   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
2450651           158             1         10.0    1.0      0.5        3.00   
2019149           155             1         57.6    1.0      0.5       12.27   
326721             88             2         22.6    1.0      0.5        0.00   
1753508            49             1         26.1    1.0      0.5        6.22   
2562731           263             2         22.6    1.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
2450651           0.0                    1.0         18.00   
2019149           0.0                    1.0         73.62   
326721            0.0                    1.0         27.60   
1753508           0.0                    1.0         37.32   
2562731           0.0                    1.0         27.60   

         congestion_surcharge  Airport_fee  hour  
2450651                   2.5         0.00     0  
2019149                   0.0         1.25     0  
326721                    2.5         0.00     0  
1753508                   2.5         0.00     0  
2562731                   2.5         0.00     0

In [61]:
df_2023_6 = pd.read_parquet('/content/2023-6.parquet')  # 6
df.info()
df_2023_6['hour'] = df_2023_6["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_6.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2341896         2  2023-04-23 00:00:35   2023-04-23 00:08:28              1.0   
2342812         2  2023-04-23 00:56:12   2023-04-23 01:00:11              1.0   
1993733         1  2023-04-20 00:34:22   2023-04-20 00:36:07              1.0   
736007          2  2023-04-08 00:00:21   2023-04-08 00:13:47              1.0   
832764          2  2023-04-09 00:42:29   2023-04-09 00:52:33              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
2341896           1.77         1.0                  N           125   
2342812           0.76         1.0                  N            79   
1993733           0.80         1.0                  N           137   
736007            2.54         1.0                  N           246   
832764            2.73         1.0                  N            74   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
2341896            68             1         10.7    1.0      0.5        3.14   
2342812           107             1          6.5    1.0      0.5        3.45   
1993733           107             1          5.8    3.5      0.5        1.00   
736007            163             1         14.9    1.0      0.5        3.98   
832764             42             2         13.5    1.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
2341896           0.0                    1.0         18.84   
2342812           0.0                    1.0         14.95   
1993733           0.0                    1.0         11.80   
736007            0.0                    1.0         23.88   
832764            0.0                    1.0         16.00   

         congestion_surcharge  Airport_fee  hour  
2341896                   2.5          0.0     0  
2342812                   2.5          0.0     0  
1993733                   2.5          0.0     0  
736007                    2.5          0.0     0  
832764                    0.0          0.0     0

In [62]:
df_2023_7 = pd.read_parquet('/content/2023-7.parquet')  # 7
df.info()
df_2023_7['hour'] = df_2023_6["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_7.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1877471         2  2023-05-17 16:43:47   2023-05-17 16:53:46              2.0   
115160          2  2023-05-02 09:59:48   2023-05-02 10:19:09              1.0   
116330          2  2023-05-02 09:44:02   2023-05-02 09:54:22              1.0   
1221386         2  2023-05-11 21:57:24   2023-05-11 22:11:12              2.0   
3235229         2  2023-05-30 12:43:29   2023-05-30 12:56:01              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1877471           1.63         1.0                  N           238   
115160            1.92         1.0                  N           142   
116330            2.10         1.0                  N           166   
1221386           1.99         1.0                  N           113   
3235229           1.82         1.0                  N           142   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1877471           236             1         11.4    2.5      0.5        2.68   
115160            140             1         17.7    0.0      0.5        4.34   
116330            239             1         12.8    0.0      0.5        3.36   
1221386           231             1         14.2    1.0      0.5        1.80   
3235229           239             2         13.5    0.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
1877471           0.0                    1.0         20.58   
115160            0.0                    1.0         26.04   
116330            0.0                    1.0         20.16   
1221386           0.0                    1.0         21.00   
3235229           0.0                    1.0         17.50   

         congestion_surcharge  Airport_fee  hour  
1877471                   2.5          0.0   0.0  
115160                    2.5          0.0   0.0  
116330                    2.5          0.0   0.0  
1221386                   2.5          0.0   0.0  
3235229                   2.5          0.0   0.0

In [63]:
df_2023_8 = pd.read_parquet('/content/2023-8.parquet')  # 8
df.info()
df_2023_8['hour'] = df_2023_8["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_8.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1784147         2  2023-07-21 00:15:20   2023-07-21 00:20:05              1.0   
2659392         2  2023-07-30 00:51:49   2023-07-30 01:18:28              1.0   
694520          2  2023-07-10 00:27:33   2023-07-10 00:40:21              1.0   
347616          1  2023-07-06 00:48:39   2023-07-06 01:01:13              1.0   
209363          2  2023-07-04 00:37:13   2023-07-04 01:09:42              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1784147           0.71         1.0                  N           144   
2659392           4.28         1.0                  N           225   
694520            2.86         1.0                  N           114   
347616            7.20         1.0                  N            88   
209363            7.28         1.0                  N           158   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1784147            79             1          6.5    1.0      0.5        1.50   
2659392           112             1         25.4    1.0      0.5        5.58   
694520            230             1         15.6    1.0      0.5        4.12   
347616            237             1         28.9    3.5      0.5        6.75   
209363            112             1         37.3    1.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
1784147           0.0                    1.0         13.00   
2659392           0.0                    1.0         33.48   
694520            0.0                    1.0         24.72   
347616            0.0                    1.0         40.65   
209363            0.0                    1.0         42.30   

         congestion_surcharge  Airport_fee  hour  
1784147                   2.5          0.0     0  
2659392                   0.0          0.0     0  
694520                    2.5          0.0     0  
347616                    2.5          0.0     0  
209363                    2.5          0.0     0

In [64]:
df_2023_9 = pd.read_parquet('/content/2023-9.parquet')  # 9
df.info()
df_2023_9['hour'] = df_2023_9["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_9.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
772202          2  2023-09-09 00:46:31   2023-09-09 00:53:56              1.0   
2158874         1  2023-09-26 00:09:48   2023-09-26 00:14:00              1.0   
1649477         2  2023-09-17 00:45:20   2023-09-17 01:08:15              6.0   
772299          2  2023-09-09 00:40:08   2023-09-09 00:45:29              1.0   
979816          2  2023-09-11 00:28:01   2023-09-11 00:42:37              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
772202            0.50         1.0                  N           186   
2158874           0.60         1.0                  N           113   
1649477           6.57         1.0                  N           239   
772299            0.87         1.0                  N           231   
979816            2.74         1.0                  N            48   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
772202            234             1          7.9    1.0      0.5        2.58   
2158874           211             1          6.5    3.5      0.5        2.30   
1649477            13             1         29.6    1.0      0.5        5.19   
772299            261             1          7.2    1.0      0.5        2.44   
979816            262             1         16.3    1.0      0.5        4.26   

         tolls_amount  improvement_surcharge  total_amount  \
772202            0.0                    1.0         15.48   
2158874           0.0                    1.0         13.80   
1649477           0.0                    1.0         39.79   
772299            0.0                    1.0         14.64   
979816            0.0                    1.0         25.56   

         congestion_surcharge  Airport_fee  hour  
772202                    2.5          0.0     0  
2158874                   2.5          0.0     0  
1649477                   2.5          0.0     0  
772299                    2.5          0.0     0  
979816                    2.5          0.0     0

In [65]:
df_2023_10 = pd.read_parquet('/content/2023-10.parquet')  # 10
df.info()
df_2023_10['hour'] = df_2023_10["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_10.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1272716         2  2023-10-13 00:18:49   2023-10-13 00:31:34              1.0   
3078075         2  2023-10-29 00:33:55   2023-10-29 00:43:25              1.0   
3467856         2  2023-10-21 00:36:47   2023-10-21 00:51:20              NaN   
1273086         2  2023-10-13 00:23:03   2023-10-13 00:34:35              2.0   
3406            2  2023-10-01 00:39:07   2023-10-01 00:42:12              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1272716           2.74         1.0                  N           231   
3078075           1.79         1.0                  N           142   
3467856           2.50         NaN               None           229   
1273086           1.95         1.0                  N           114   
3406              0.51         1.0                  N           230   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1272716           224             1        14.90    1.0      0.5        3.98   
3078075            68             1        11.40    1.0      0.5        3.28   
3467856            79             0        18.21    0.0      0.5        2.00   
1273086            13             1        12.80    1.0      0.5        3.56   
3406              186             1         5.10    1.0      0.5        0.01   

         tolls_amount  improvement_surcharge  total_amount  \
1272716           0.0                    1.0         23.88   
3078075           0.0                    1.0         19.68   
3467856           0.0                    1.0         24.21   
1273086           0.0                    1.0         21.36   
3406              0.0                    1.0         10.11   

         congestion_surcharge  Airport_fee  hour  
1272716                   2.5          0.0     0  
3078075                   2.5          0.0     0  
3467856                   NaN          NaN     0  
1273086                   2.5          0.0     0  
3406                      2.5          0.0     0

In [66]:
df_2023_11 = pd.read_parquet('/content/2023-11.parquet')  # 11
df.info()
df_2023_11['hour'] = df_2023_11["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_11.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1035965         2  2023-11-10 00:32:24   2023-11-10 00:43:58              1.0   
1037547         1  2023-11-10 00:15:29   2023-11-10 00:17:40              1.0   
2078188         2  2023-11-19 00:39:47   2023-11-19 00:44:02              1.0   
2600252         2  2023-11-25 00:37:20   2023-11-25 00:43:57              1.0   
3088309         2  2023-11-30 00:10:50   2023-11-30 00:16:10              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1035965           2.06         1.0                  N           186   
1037547           0.30         1.0                  N            48   
2078188           0.66         1.0                  N           141   
2600252           1.50         1.0                  N            48   
3088309           0.94         1.0                  N           230   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1035965           229             2         13.5    1.0      0.5        0.00   
1037547            48             1          4.4    3.5      0.5        0.09   
2078188           263             2          5.8    1.0      0.5        0.00   
2600252           143             1          9.3    1.0      0.5        2.86   
3088309            68             2          7.2    1.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
1035965           0.0                    1.0         18.50   
1037547           0.0                    1.0          9.49   
2078188           0.0                    1.0         10.80   
2600252           0.0                    1.0         17.16   
3088309           0.0                    1.0         12.20   

         congestion_surcharge  Airport_fee  hour  
1035965                   2.5          0.0     0  
1037547                   2.5          0.0     0  
2078188                   2.5          0.0     0  
2600252                   2.5          0.0     0  
3088309                   2.5          0.0     0

In [67]:
df_2023_12 = pd.read_parquet('/content/2023-12.parquet')  # 12
df.info()
df_2023_12['hour'] = df_2023_12["tpep_pickup_datetime"].dt.hour
sample_df_2023 = df_2023_12.groupby("hour").sample(frac=0.05, random_state=42)
import os #Navigate Google Drive
path = "/content/drive/My Drive/Colab Notebooks/Assignments"
os.makedirs(path, exist_ok=True)
sample_df_2023.to_parquet("/content/drive/My Drive/Colab Notebooks/Assignments/nyc_taxi_sample1.parquet",index=False) # address of
sample_df_2023.info()
sample_df_2023.head()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2225435         2  2023-12-20 00:19:54   2023-12-20 00:30:46              1.0   
1916769         1  2023-12-17 00:48:57   2023-12-17 00:51:54              1.0   
1920448         1  2023-12-17 00:05:01   2023-12-17 00:14:14              2.0   
3306208         2  2023-12-17 00:24:42   2023-12-17 00:46:16              NaN   
2800410         1  2023-12-27 00:58:36   2023-12-27 01:04:13              1.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
2225435           2.55         1.0                  N           143   
1916769           0.80         1.0                  N           170   
1920448           1.20         1.0                  N            68   
3306208           3.01         NaN               None            79   
2800410           2.00         1.0                  N            79   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
2225435            68             1         14.9    1.0      0.5        3.98   
1916769           233             1          5.8    3.5      0.5        3.20   
1920448            68             1         10.0    3.5      0.5        3.00   
3306208           237             0         24.7    0.0      0.5        5.74   
2800410           170             1         10.0    3.5      0.5        3.75   

         tolls_amount  improvement_surcharge  total_amount  \
2225435           0.0                    1.0         23.88   
1916769           0.0                    1.0         14.00   
1920448           0.0                    1.0         18.00   
3306208           0.0                    1.0         34.44   
2800410           0.0                    1.0         18.75   

         congestion_surcharge  Airport_fee  hour  
2225435                   2.5          0.0     0  
1916769                   2.5          0.0     0  
1920448                   2.5          0.0     0  
3306208                   NaN          NaN     0  
2800410                   2.5          0.0     0

**1.1.1** <font color = red>[5 marks]</font> <br>
Figure out how to sample and combine the files.

**Note:** It is not mandatory to use the method specified above. While sampling, you only need to make sure that your sampled data represents the overall data of all the months accurately.

In [ ]:
# Sample the data
# It is recommmended to not load all the files at once to avoid memory overload

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

In [ ]:
# Take a small percentage of entries from each hour of every date.
# Iterating through the monthly data:
#   read a month file -> day -> hour: append sampled data -> move to next hour -> move to next day after 24 hours -> move to next month file
# Create a single dataframe for the year combining all the monthly data

# Select the folder having data files
import os

# Select the folder having data files
os.chdir('/content/Assignments/EDA/data_NYC_Taxi/trip_records')

# Create a list of all the twelve files to read
file_list = os.listdir()

# initialise an empty dataframe
df = pd.DataFrame()


# iterate through the list of files and sample one by one:
for file_name in file_list:
    try:
        # file path for the current file
        file_path = os.path.join(os.getcwd(), file_name)

        # Reading the current file



        # We will store the sampled data for the current date in this df by appending the sampled data from each hour to this
        # After completing iteration through each date, we will append this data to the final dataframe.
        sampled_data = pd.DataFrame()

        # Loop through dates and then loop through every hour of each date

            # Iterate through each hour of the selected date

                # Sample 5% of the hourly data randomly

                # add data of this hour to the dataframe

        # Concatenate the sampled data of all the dates to a single dataframe
        df = # we initialised this empty DF earlier

    except Exception as e:
        print(f"Error reading file {file_name}: {e}")

After combining the data files into one DataFrame, convert the new DataFrame to a CSV or parquet file and store it to use directly.

Ideally, you can try keeping the total entries to around 250,000 to 300,000.

In [ ]:
# Store the df in csv/parquet
# df.to_parquet('')

## **2** Data Cleaning
<font color = red>[30 marks]</font> <br>

Now we can load the new data directly.

In [ ]:
# Load the new data file



In [ ]:
# df.head()

In [ ]:
# df.info()

#### **2.1** Fixing Columns
<font color = red>[10 marks]</font> <br>

Fix/drop any columns as you seem necessary in the below sections

**2.1.1** <font color = red>[2 marks]</font> <br>

Fix the index and drop unnecessary columns

In [ ]:
# Fix the index and drop any columns that are not needed



**2.1.2** <font color = red>[3 marks]</font> <br>
There are two airport fee columns. This is possibly an error in naming columns. Let's see whether these can be combined into a single column.

In [ ]:
# Combine the two airport fee columns



**2.1.3** <font color = red>[5 marks]</font> <br>
Fix columns with negative (monetary) values

In [ ]:
# check where values of fare amount are negative



Did you notice something different in the `RatecodeID` column for above records?

In [ ]:
# Analyse RatecodeID for the negative fare amounts



In [ ]:
# Find which columns have negative values



In [ ]:
# fix these negative values



### **2.2** Handling Missing Values
<font color = red>[10 marks]</font> <br>

**2.2.1**  <font color = red>[2 marks]</font> <br>
Find the proportion of missing values in each column




In [ ]:
# Find the proportion of missing values in each column



**2.2.2**  <font color = red>[3 marks]</font> <br>
Handling missing values in `passenger_count`

In [ ]:
# Display the rows with null values
# Impute NaN values in 'passenger_count'


Did you find zeroes in passenger_count? Handle these.

**2.2.3**  <font color = red>[2 marks]</font> <br>
Handle missing values in `RatecodeID`

In [ ]:
# Fix missing values in 'RatecodeID'


**2.2.4**  <font color = red>[3 marks]</font> <br>
Impute NaN in `congestion_surcharge`

In [ ]:
# handle null values in congestion_surcharge




Are there missing values in other columns? Did you find NaN values in some other set of columns? Handle those missing values below.

In [ ]:
# Handle any remaining missing values



### **2.3** Handling Outliers
<font color = red>[10 marks]</font> <br>

Before we start fixing outliers, let's perform outlier analysis.

In [ ]:
# Describe the data and check if there are any potential outliers present
# Check for potential out of place values in various columns



**2.3.1**  <font color = red>[10 marks]</font> <br>
Based on the above analysis, it seems that some of the outliers are present due to errors in registering the trips. Fix the outliers.

Some points you can look for:
- Entries where `trip_distance` is nearly 0 and `fare_amount` is more than 300
- Entries where `trip_distance` and `fare_amount` are 0 but the pickup and dropoff zones are different (both distance and fare should not be zero for different zones)
- Entries where `trip_distance` is more than 250  miles.
- Entries where `payment_type` is 0 (there is no payment_type 0 defined in the data dictionary)

These are just some suggestions. You can handle outliers in any way you wish, using the insights from above outlier analysis.

How will you fix each of these values? Which ones will you drop and which ones will you replace?

First, let us remove 7+ passenger counts as there are very less instances.

In [ ]:
# remove passenger_count > 6


In [ ]:
# Continue with outlier handling



In [ ]:
# Do any columns need standardising?



## **3** Exploratory Data Analysis
<font color = red>[90 marks]</font> <br>

In [ ]:
df.columns.tolist()

#### **3.1** General EDA: Finding Patterns and Trends
<font color = red>[40 marks]</font> <br>

**3.1.1** <font color = red>[3 marks]</font> <br>
Categorise the varaibles into Numerical or Categorical.
* `VendorID`:
* `tpep_pickup_datetime`:
* `tpep_dropoff_datetime`:
* `passenger_count`:
* `trip_distance`:
* `RatecodeID`:
* `PULocationID`:
* `DOLocationID`:
* `payment_type`:
* `pickup_hour`:
* `trip_duration`:


The following monetary parameters belong in the same category, is it categorical or numerical?


* `fare_amount`
* `extra`
* `mta_tax`
* `tip_amount`
* `tolls_amount`
* `improvement_surcharge`
* `total_amount`
* `congestion_surcharge`
* `airport_fee`

##### Temporal Analysis

**3.1.2** <font color = red>[5 marks]</font> <br>
Analyse the distribution of taxi pickups by hours, days of the week, and months.

In [ ]:
# Find and show the hourly trends in taxi pickups



In [ ]:
# Find and show the daily trends in taxi pickups (days of the week)



In [ ]:
# Show the monthly trends in pickups



##### Financial Analysis

Take a look at the financial parameters like `fare_amount`, `tip_amount`, `total_amount`, and also `trip_distance`. Do these contain zero/negative values?

In [ ]:
# Analyse the above parameters



Do you think it is beneficial to create a copy DataFrame leaving out the zero values from these?

**3.1.3** <font color = red>[2 marks]</font> <br>
Filter out the zero values from the above columns.

**Note:** The distance might be 0 in cases where pickup and drop is in the same zone. Do you think it is suitable to drop such cases of zero distance?

In [ ]:
# Create a df with non zero entries for the selected parameters.



**3.1.4** <font color = red>[3 marks]</font> <br>
Analyse the monthly revenue (`total_amount`) trend

In [ ]:
# Group data by month and analyse monthly revenue



**3.1.5** <font color = red>[3 marks]</font> <br>
Show the proportion of each quarter of the year in the revenue

In [ ]:
# Calculate proportion of each quarter



**3.1.6** <font color = red>[3 marks]</font> <br>
Visualise the relationship between `trip_distance` and `fare_amount`. Also find the correlation value for these two.

**Hint:** You can leave out the trips with trip_distance = 0

In [ ]:
# Show how trip fare is affected by distance



**3.1.7** <font color = red>[5 marks]</font> <br>
Find and visualise the correlation between:
1. `fare_amount` and trip duration (pickup time to dropoff time)
2. `fare_amount` and `passenger_count`
3. `tip_amount` and `trip_distance`

In [ ]:
# Show relationship between fare and trip duration



In [ ]:
# Show relationship between fare and number of passengers



In [ ]:
# Show relationship between tip and trip distance



**3.1.8** <font color = red>[3 marks]</font> <br>
Analyse the distribution of different payment types (`payment_type`)

In [ ]:
# Analyse the distribution of different payment types (payment_type).




- 1= Credit card
- 2= Cash
- 3= No charge
- 4= Dispute



##### Geographical Analysis

For this, you have to use the *taxi_zones.shp* file from the *taxi_zones* folder.

There would be multiple files inside the folder (such as *.shx, .sbx, .sbn* etc). You do not need to import/read any of the files other than the shapefile, *taxi_zones.shp*.

Do not change any folder structure - all the files need to be present inside the folder for it to work.

The folder structure should look like this:
```
Taxi Zones
|- taxi_zones.shp.xml
|- taxi_zones.prj
|- taxi_zones.sbn
|- taxi_zones.shp
|- taxi_zones.dbf
|- taxi_zones.shx
|- taxi_zones.sbx

 ```

 You only need to read the `taxi_zones.shp` file. The *shp* file will utilise the other files by itself.

We will use the *GeoPandas* library for geopgraphical analysis
```
import geopandas as gpd
```

More about geopandas and shapefiles: [About](https://geopandas.org/en/stable/about.html)


Reading the shapefile is very similar to *Pandas*. Use `gpd.read_file()` function to load the data (*taxi_zones.shp*) as a GeoDataFrame. Documentation: [Reading and Writing Files](https://geopandas.org/en/stable/docs/user_guide/io.html)

In [ ]:
# !pip install geopandas

**3.1.9** <font color = red>[2 marks]</font> <br>
Load the shapefile and display it.

In [ ]:
# import geopandas as gpd


# Read the shapefile using geopandas
zones = # read the .shp file using gpd
zones.head()

Now, if you look at the DataFrame created, you will see columns like: `OBJECTID`,`Shape_Leng`, `Shape_Area`, `zone`, `LocationID`, `borough`, `geometry`.
<br><br>

Now, the `locationID` here is also what we are using to mark pickup and drop zones in the trip records.

The geometric parameters like shape length, shape area and geometry are used to plot the zones on a map.

This can be easily done using the `plot()` method.

In [ ]:
# print(zones.info())
# zones.plot()

Now, you have to merge the trip records and zones data using the location IDs.



**3.1.10** <font color = red>[3 marks]</font> <br>
Merge the zones data into trip data using the `locationID` and `PULocationID` columns.

In [ ]:
# Merge zones and trip records using locationID and PULocationID



**3.1.11** <font color = red>[3 marks]</font> <br>
Group data by location IDs to find the total number of trips per location ID

In [ ]:
# Group data by location and calculate the number of trips



**3.1.12** <font color = red>[2 marks]</font> <br>
Now, use the grouped data to add number of trips to the GeoDataFrame.

We will use this to plot a map of zones showing total trips per zone.

In [ ]:
# Merge trip counts back to the zones GeoDataFrame




The next step is creating a color map (choropleth map) showing zones by the number of trips taken.

Again, you can use the `zones.plot()` method for this. [Plot Method GPD](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html#geopandas.GeoDataFrame.plot)

But first, you need to define the figure and axis for the plot.

`fig, ax = plt.subplots(1, 1, figsize = (12, 10))`

This function creates a figure (fig) and a single subplot (ax)

---

After setting up the figure and axis, we can proceed to plot the GeoDataFrame on this axis. This is done in the next step where we use the plot method of the GeoDataFrame.

You can define the following parameters in the `zones.plot()` method:
```
column = '',
ax = ax,
legend = True,
legend_kwds = {'label': "label", 'orientation': "<horizontal/vertical>"}
```

To display the plot, use `plt.show()`.

**3.1.13** <font color = red>[3 marks]</font> <br>
Plot a color-coded map showing zone-wise trips

In [ ]:
# Define figure and axis


# Plot the map and display it



In [ ]:
# can you try displaying the zones DF sorted by the number of trips?



Here we have completed the temporal, financial and geographical analysis on the trip records.

**Compile your findings from general analysis below:**

You can consider the following points:

* Busiest hours, days and months
* Trends in revenue collected
* Trends in quarterly revenue
* How fare depends on trip distance, trip duration and passenger counts
* How tip amount depends on trip distance
* Busiest zones


#### **3.2** Detailed EDA: Insights and Strategies
<font color = red>[50 marks]</font> <br>

Having performed basic analyses for finding trends and patterns, we will now move on to some detailed analysis focussed on operational efficiency, pricing strategies, and customer experience.

##### Operational Efficiency

Analyze variations by time of day and location to identify bottlenecks or inefficiencies in routes

**3.2.1** <font color = red>[3 marks]</font> <br>
Identify slow routes by calculating the average time taken by cabs to get from one zone to another at different hours of the day.

Speed on a route *X* for hour *Y* = (*distance of the route X / average trip duration for hour Y*)

In [ ]:
# Find routes which have the slowest speeds at different times of the day



How does identifying high-traffic, high-demand routes help us?

**3.2.2** <font color = red>[3 marks]</font> <br>
Calculate the number of trips at each hour of the day and visualise them. Find the busiest hour and show the number of trips for that hour.

In [ ]:
# Visualise the number of trips per hour and find the busiest hour



Remember, we took a fraction of trips. To find the actual number, you have to scale the number up by the sampling ratio.

**3.2.3** <font color = red>[2 mark]</font> <br>
Find the actual number of trips in the five busiest hours

In [ ]:
# Scale up the number of trips

# Fill in the value of your sampling fraction and use that to scale up the numbers
sample_fraction =



SyntaxError: invalid syntax (ipython-input-2793492136.py, line 4)

**3.2.4** <font color = red>[3 marks]</font> <br>
Compare hourly traffic pattern on weekdays. Also compare for weekend.

In [ ]:
# Compare traffic trends for the week days and weekends



What can you infer from the above patterns? How will finding busy and quiet hours for each day help us?

**3.2.5** <font color = red>[3 marks]</font> <br>
Identify top 10 zones with high hourly pickups. Do the same for hourly dropoffs. Show pickup and dropoff trends in these zones.

In [ ]:
# Find top 10 pickup and dropoff zones



**3.2.6** <font color = red>[3 marks]</font> <br>
Find the ratio of pickups and dropoffs in each zone. Display the 10 highest (pickup/drop) and 10 lowest (pickup/drop) ratios.

In [ ]:
# Find the top 10 and bottom 10 pickup/dropoff ratios



**3.2.7** <font color = red>[3 marks]</font> <br>
Identify zones with high pickup and dropoff traffic during night hours (11PM to 5AM)

In [ ]:
# During night hours (11pm to 5am) find the top 10 pickup and dropoff zones
# Note that the top zones should be of night hours and not the overall top zones



Now, let us find the revenue share for the night time hours and the day time hours. After this, we will move to deciding a pricing strategy.

**3.2.8** <font color = red>[2 marks]</font> <br>
Find the revenue share for nighttime and daytime hours.

In [ ]:
# Filter for night hours (11 PM to 5 AM)



##### Pricing Strategy

**3.2.9** <font color = red>[2 marks]</font> <br>
For the different passenger counts, find the average fare per mile per passenger.

For instance, suppose the average fare per mile for trips with 3 passengers is 3 USD/mile, then the fare per mile per passenger will be 1 USD/mile.

In [ ]:
# Analyse the fare per mile per passenger for different passenger counts




**3.2.10** <font color = red>[3 marks]</font> <br>
Find the average fare per mile by hours of the day and by days of the week

In [ ]:
# Compare the average fare per mile for different days and for different times of the day



**3.2.11** <font color = red>[3 marks]</font> <br>
Analyse the average fare per mile for the different vendors for different hours of the day

In [ ]:
# Compare fare per mile for different vendors



**3.2.12** <font color = red>[5 marks]</font> <br>
Compare the fare rates of the different vendors in a tiered fashion. Analyse the average fare per mile for distances upto 2 miles. Analyse the fare per mile for distances from 2 to 5 miles. And then for distances more than 5 miles.


In [ ]:
# Defining distance tiers



##### Customer Experience and Other Factors

**3.2.13** <font color = red>[5 marks]</font> <br>
Analyse average tip percentages based on trip distances, passenger counts and time of pickup. What factors lead to low tip percentages?

In [ ]:
#  Analyze tip percentages based on distances, passenger counts and pickup times



Additional analysis [optional]: Let's try comparing cases of low tips with cases of high tips to find out if we find a clear aspect that drives up the tipping behaviours

In [ ]:
# Compare trips with tip percentage < 10% to trips with tip percentage > 25%



**3.2.14** <font color = red>[3 marks]</font> <br>
Analyse the variation of passenger count across hours and days of the week.

In [ ]:
# See how passenger count varies across hours and days




**3.2.15** <font color = red>[2 marks]</font> <br>
Analyse the variation of passenger counts across zones

In [ ]:
# How does passenger count vary across zones



In [ ]:
# For a more detailed analysis, we can use the zones_with_trips GeoDataFrame
# Create a new column for the average passenger count in each zone.



Find out how often surcharges/extra charges are applied to understand their prevalance

**3.2.16** <font color = red>[5 marks]</font> <br>
Analyse the pickup/dropoff zones or times when extra charges are applied more frequently

In [ ]:
# How often is each surcharge applied?



## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.